### ISSUE TITLE
Reproducing BLIP3o 4B model from scratch

### ISSUE THREAD

**Issue by mmderakhshani** [JiuhaiChen/BLIP3o#27]
Hi @JiuhaiChen,

Thank you very much for releasing the base code and dataset. I would like to try this repo and reproduce your 4B model from scratch on 256 GPUs. 

1. Could you please let me know the exact slurm? 
2. Which settings did you use for both pretraining and instruction tuning? 

I am asking this because I did not see the `slurm.sh` file for the 4B model's pretraining and fine-tuning, and on the other hand, I would like to make sure that, in terms of hyperparameters and datasets, I am not deviating too much. Thank you. 

Best,
Mohammad

**Comment by JiuhaiChen** [comment]
Hi, thanks for your interest!

**Pretraining the 4B model**

1. Use the `slurm.sh` script ([https://github.com/JiuhaiChen/BLIP3o/blob/main/slurm.sh](https://github.com/JiuhaiChen/BLIP3o/blob/main/slurm.sh)), setting `--model_name_or_path` to `Qwen/Qwen2.5-VL-3B-Instruct`.
2. Download all three released datasets—Long Caption, Short Caption, and JourneyDB, see https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L587 or  Training section in README, to load all these three datasets.
3. If you’re training the diffusion transformer from scratch, we suggest at least **150 K steps** (roughly 10 epochs).

**Instruction tuning**

We just uploaded a sft.sh for your reference, change `--model_name_or_path` to your pretrained 4B checkpoint and set the dataset path.

You can use our **BLIP3o-60k** instruction tuning dataset and we suggest run at least **10 K steps** for best results.


Let me know if you have any questions.

**Comment by mmderakhshani** [comment]
Thank you very much for your quick reply. I will keep you posted and release my reproduction's logs/artifacts here.

**Comment by mmderakhshani** [comment]
Now I have set up everything. 

Could you please let me know whether the current `slurm.sh` does training of the diffusion transformers from scratch or not? And if yes, could you please let me know how I can quickly test the pretrained version of the diffusion transformer? 

For the 4B version, you trained for 10 epochs, right?

**Comment by JiuhaiChen** [comment]
Hi, for the current slurm.sh, it is training the diffusion transformer from scratch. For the quickly test, you can try if the model can generate reasonable images. You can use the same inference code provided in our repo, the only thing you need to import is the diffusion decoder, which can be found here:

https://huggingface.co/BLIP3o/BLIP3o-Model-8B/tree/main/diffusion-decoder

**Comment by mmderakhshani** [comment]
Perfect, will get back to you whenever I have results. Thank you.

**Comment by mmderakhshani** [comment]
> If you’re training the diffusion transformer from scratch, we suggest at least 150 K steps (roughly 10 epochs).

I am a bit confused about your previous comment. Regarding the above point, what should I actually change? Do you mean I need to change the number of epochs from 1 to 10 at this [line](https://github.com/JiuhaiChen/BLIP3o/blob/main/slurm.sh#L29) ?  What do you mean by diffusion model? Is this `lumina-Next`?

**Comment by mmderakhshani** [comment]
Perfect. Thanks for this. Can you tell me exactly how long it took you for the full run of the 4B model pretraining? 

What about this part:
> For the quickly test, you can try if the model can generate reasonable images. You can use the same inference code provided in our repo, the only thing you need to import is the diffusion decoder, which can be found here: https://huggingface.co/BLIP3o/BLIP3o-Model-8B/tree/main/diffusion-decoder

How should I change the slurm.sh script. Thank you in advance.

**Comment by JiuhaiChen** [comment]
Hi, it takes 10 days to run 4B pretraining on 32 H200. 

This part is inference part, you need to import your model path. And update the diffusion path: https://github.com/JiuhaiChen/BLIP3o/blob/main/inference.py#L32
The diffusion decoder can be found from https://huggingface.co/BLIP3o/BLIP3o-Model-8B/tree/main/diffusion-decoder.


Slurm.sh you only need to change the data path, save model path and some slurm config.

**Comment by fengyang0317** [comment]
Hi @JiuhaiChen ,

Thanks for the great work. I have a question regarding the training settings. I get 38,350,681 examples after combining the three datasets. Using 32 gpus with batch_size 16 per-device, the global batch size will be 512. 150k steps * 512 batch_size = 76,800,000 exampels, which is roughly 2 epochs. 

As you mentioned, you trained 10 epochs. Did you increase the number of devices or batch_size per-device? Or did yo trained for about 750k steps?

**Comment by JiuhaiChen** [comment]
Hi, thanks for your interest! On our H200 GPUs (140 GB memory) we use a batch size of 32. We recommend pushing the batch size as high as your GPU allows and training for at least 150 k steps, performance generally improves with more steps. The 10-epoch setting is only a predefine value, we didn’t end up completing all 10 epochs. You can stopping any time when you feel the model performance is acceptable.

**Comment by mmderakhshani** [comment]
Hey, quick question — after finishing the pretraining stage of my 4B model, I want to start instruction tuning using the weights from the last checkpoint, but without carrying over the optimizer, scheduler, or training progress (like global steps).

What’s the proper way to do this? Is it enough to load the model with from_pretrained() and skip resume_from_checkpoint? Or is there a more standard approach?

I’m using your code, and when I run sft.sh, it loads the previous stage’s checkpoint — but during training, it continues from the last global step of the pretraining phase. I’d like instruction tuning to start from step zero instead.

**Comment by JiuhaiChen** [comment]
Hi, we didn't using any optimizer and scheduler from pretrain when doing the instruction tuning. You can use the sft.sh and change the model_name_or_path with your pretrain model name, and also update the data folder.

You can remove the checkpoint folder under your pretrain model folder in case the training will resume from it.

**Comment by mmderakhshani** [comment]
Great, resolved! Can you please tell me how many iterations you train the BLIP3o4 B model on the instruction tuning dataset? For pretraining, I remembered you mentioned 150k. What about instruction tuning?

**Comment by JiuhaiChen** [comment]
Hi, thanks for your interest! For DPG score, can you reproduce the other model's score? Like Janus-pro?

**Comment by mmderakhshani** [comment]
Unfortunately not. I could not. So if you can run DPG eval script on my
generated image, then this means that my there is something missing in my
eval pipeline. That is a very big favor to me honestly as I could connect
all pieces together.

PhD student at AIM Lab
Science Park, University of Amsterdam, Amsterdam, Netherlands
Skype Id: m2.derakhshani1993

**Comment by JiuhaiChen** [comment]
Sure, will get back to you soon.

**Comment by mmderakhshani** [comment]
Thank you very much @JiuhaiChen. I am looking forward to hearing back from you soon.

**Comment by mmderakhshani** [comment]
Dear @JiuhaiChen, is there any update on computing the DPG score?
